src/marine/argoClim2ioda.py.in

```diff
@@ -0,0 +1,180 @@
+        '''
+
+        try:
+            nc = nc4.Dataset(self.filename, 'r')
+        except IOError:
+            raise IOError('%s file not found!' % self.filename)
+        except Exception:
+            raise Exception('Unknown error opening %s' % self.filename)
+
+        # First find the variable in this file
+        res = [x for x in list(nc.variables.keys()) if re.search("^ARGO_.*MEAN$", x)]
+
+        if res:
+            self.varname = res[0].split('_')[1]
+            print("%s contains the variable: %s" % (self.filename, self.varname))
+        else:
+            raise NameError("No useable variable was found in the file %s" % self.filename)
+
+        assert self.varname in ['TEMPERATURE', 'SALINITY'],\
+            "%s is not a valid variable name" % self.varname
+
+        lon = nc.variables['LONGITUDE'][:]
+        lat = nc.variables['LATITUDE'][:]
+        pres = nc.variables['PRESSURE'][:]
+
+        anomaly = nc.variables['ARGO_%s_ANOMALY' % self.varname][:]
+        mean = nc.variables['ARGO_%s_MEAN' % self.varname][:]
+
+        # Get absolute time instead of time since epoch
+        dtime = nc.variables['TIME']
+        time = nc4.num2date(dtime[:], dtime.units, calendar='360_day')
+
+        # create a full field from mean and anomaly
+        fullField = anomaly + np.tile(mean, (anomaly.shape[0], 1, 1, 1))
+
+        try:
+            nc.close()
+        except IOError:
+            raise IOError('%s file could not be closed!' % self.filename)
+        except Exception:
+            raise Exception('Unknown error closing %s' % self.filename)
+
+        # self.data is the data structure
+        self.data = {}
+        self.data['lat'] = lat
+        self.data['lon'] = lon
+        self.data['pres'] = pres
+        self.data['time'] = time
+        self.data['field'] = fullField.data
+
+        return
+
+
+class IODA(object):
+
+    def __init__(self, filename, date, argo):
+        '''
+        Initialize IODA writer class,
+        transform to IODA data structure and,
+        write out to IODA file.
+        '''
+
+        self.filename = filename
+        self.date = date
+
+        self.locKeyList = [
+            ("latitude", "float"),
+            ("longitude", "float"),
+            ("pressure", "float"),
+            ("datetime", "string")
+        ]
+
+        self.AttrData = {
+            'odb_version': 1,
+            'date_time_string': self.date.strftime("%Y-%m-%dT%H:%M:%SZ")
+        }
+
+        self.writer = iconv.NcWriter(self.filename, [], self.locKeyList)
+
+        # data is the dictionary containing IODA friendly data structure
+        self.data = DefaultOrderedDict(lambda: DefaultOrderedDict(dict))
+
+        recKey = 0
+        valKey = argo.varname, self.writer.OvalName()
+        errKey = argo.varname, self.writer.OerrName()
+        qcKey = argo.varname, self.writer.OqcName()
+
+        # There has to be a better way than explicitly looping over 4! dimensions
```

Comment:
@srherbener 
Any suggestions here.
This code block is painfully slow.
Imagine approx. `ntime = 200, npres = 60, nlat = 180 , nlon = 360`

Reply:
@aerorahul, the NcWriter methods got refactored to provide two flows.

The first flow is intended for data that is organized as a list of observations where each item in the list has the obs value and all metadata, and you need to group the obs that have identical locations together and fill in the missing values as you go. Conventional data tends to be like this (radiosonde sounding, eg).

The second flow is intended for data that is already organized by locations (perhaps in an array) and either doesn't have missing data, or the missing data is provided (so there is no need to find where the missing data is). Radiance data tends to be like this.

The first flow is to do as you did below, and created the multi-key dictionary with the rec, loc, and val keys and then call ExtractObsData and BuildNetcdf.

The second flow is to do as @travissluka is suggesting which is to bypass the ExtractObsData step by building the ObsVars and the Rec, Loc, Var, Attr Data dictionaries in your script and only call BuildNetcdf.

The building of the multi-keyed dictionary and the execution of ExtractObsData is inefficient (slow), but it was assumed that this would primarily be conventional data that would not get excessively large. (Perhaps this is not a good assumption?) This code could likely be improved too, but we haven't had time to address that yet.

It looks like your input data (argoclim_test.nc) is already organized by location (x,y,z,t or subset), so I think @travissluka's suggestion is the best approach. It seems you should be able to read the arrays in and treat them as vectors when writing them out. You would have to create the corresponding time, lat, lon ,press vectors but perhaps this could be done by repeating blocks instead of walking through the 4-level-deep for loops. In any case, avoiding the call to ExtractObsData is a big time saver.

I'm happy to help further if the above suggestion doesn't work out.